File: app/Models/Subject.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Subject extends Model
{
    protected $fillable = [
        'name', 'name_ar', 'description', 'parent_id', 'is_main', 'is_religion_based', 'religion'
    ];

    protected function casts(): array
    {
        return [
            'is_main' => 'boolean',
            'is_religion_based' => 'boolean',
        ];
    }

    public function parent()
    {
        return $this->belongsTo(Subject::class, 'parent_id');
    }

    public function children()
    {
        return $this->hasMany(Subject::class, 'parent_id');
    }

    public function grades()
    {
        return $this->belongsToMany(Grade::class, 'grade_subject');
    }

    public function scopeRoot($query)
    {
        return $query->whereNull('parent_id');
    }

    public function scopeMain($query)
    {
        return $query->where('is_main', true);
    }

    public function scopeOptional($query)
    {
        return $query->where('is_main', false);
    }

    public function scopeForReligion($query, ?string $religion)
    {
        if (!$religion) return $query->where('is_religion_based', false);

        return $query->where(function ($q) use ($religion) {
            $q->where('is_religion_based', false)
              ->orWhere(function ($sub) use ($religion) {
                  $sub->where('is_religion_based', true)
                      ->where('religion', $religion);
              });
        });
    }
}

File: flutter/lib/models/subject.dart
class Subject {
  final int id;
  final String name;
  final String nameAr;
  final String? description;
  final int? parentId;
  final bool isMain;
  final bool isReligionBased;
  final String? religion;
  final List<Subject>? children;

  Subject({
    required this.id,
    required this.name,
    required this.nameAr,
    this.description,
    this.parentId,
    this.isMain = true,
    this.isReligionBased = false,
    this.religion,
    this.children,
  });

  factory Subject.fromJson(Map<String, dynamic> json) => Subject(
    id: json['id'],
    name: json['name'] ?? '',
    nameAr: json['name_ar'] ?? '',
    description: json['description'],
    parentId: json['parent_id'],
    isMain: json['is_main'] ?? true,
    isReligionBased: json['is_religion_based'] ?? false,
    religion: json['religion'],
    children: json['children'] != null
        ? (json['children'] as List).map((e) => Subject.fromJson(e)).toList()
        : null,
  );

  Map<String, dynamic> toJson() => {
    'name': name,
    'name_ar': nameAr,
    'description': description,
    'parent_id': parentId,
    'is_main': isMain,
    'is_religion_based': isReligionBased,
    'religion': religion,
  };
}
